src/OrchardCore/OrchardCore.Setup.Core/UrlService.cs

```diff
@@ -0,0 +1,50 @@
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using Microsoft.AspNetCore.Http;
+using OrchardCore.Environment.Shell;
+
+namespace OrchardCore.Setup.Core
+{
+    public class UrlService : IUrlService
+    {
+        private readonly IHttpContextAccessor _httpContextAccessor;
+        public UrlService(IHttpContextAccessor httpContextAccessor)
+        {
+            _httpContextAccessor = httpContextAccessor;
+        }
+        public string GetEncodedUrl(ShellSettings shellSettings, Dictionary<string, string> queryParams = null)
+        {
+            string baseUrl = string.Empty;            
+            HostString hostString;
+            var httpContext = _httpContextAccessor.HttpContext;
+
+            var tenantUrlHost = shellSettings.RequestUrlHost?.Split(new[] { ',', ' ' }, StringSplitOptions.RemoveEmptyEntries).FirstOrDefault();
+
+            if (string.IsNullOrEmpty(tenantUrlHost))
+            {
+                hostString = httpContext.Request.Host;
+            }
+            else
+            {
+                hostString = new HostString(tenantUrlHost);
+            }
+
+            var pathString = httpContext.Features.Get<ShellContextFeature>().OriginalPathBase;
+            if (!String.IsNullOrEmpty(shellSettings.RequestUrlPrefix))
+            {
+                pathString = pathString.Add('/' + shellSettings.RequestUrlPrefix);
+            }
+
+            baseUrl = $"{httpContext.Request.Scheme}://{hostString.ToUriComponent() + pathString.ToUriComponent()}";
```

Comment:
For info here no need to have `ToUriComponent()` (that url encodes the related uri component), here the `ToString()` is implicitly called, and it calls `ToUriComponent()`.

Then i like to have in one line at the end what aspnetcore also uses, but not always ;)

            $"{Request.Scheme}://{hostString + pathString + queryString}";

Then, to build a non encoded url for display only, you have to use the `.Value` property of each component, which is the original unencoded (unescaped) string.

Unless for a `QueryString` component whose ctor needs an already escaped string ;) So not well adapted for display only

        /// Initialize the query string with a given value. This value must be in escaped and delimited format with
        /// a leading '?' character.
        /// </summary>
        /// <param name="value">The query string to be assigned to the Value property.</param>
        public QueryString(string? value)
        {
            if (!string.IsNullOrEmpty(value) && value[0] != '?')
            {
                throw new ArgumentException("The leading '?' must be included for a non-empty query.", nameof(value));
            }
            Value = value;
        }